File: services/sabwa-engine/src/db/groups.rs
//! `sabwa_groups` — group-specific metadata cache.
//!
//! See SABWA_PLAN.md §3 "sabwa_groups".

use anyhow::{Context, Result};
use bson::{doc, oid::ObjectId, Bson};
use chrono::{DateTime, Utc};
use futures::TryStreamExt;
use mongodb::{Collection, Database};
use serde::{Deserialize, Serialize};

pub const COLLECTION: &str = "sabwa_groups";

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Participant {
    pub jid: String,
    #[serde(default)]
    pub is_admin: bool,
    #[serde(default)]
    pub is_super_admin: bool,
    pub joined_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SabwaGroup {
    #[serde(rename = "_id", skip_serializing_if = "Option::is_none")]
    pub id: Option<ObjectId>,
    pub project_id: ObjectId,
    pub session_id: ObjectId,
    /// Always ends in `@g.us`.
    pub jid: String,
    pub subject: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub description: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub creator: Option<String>,
    pub created_at: DateTime<Utc>,
    #[serde(default)]
    pub participants: Vec<Participant>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub invite_code: Option<String>,
    #[serde(default)]
    pub announcement: bool,
    #[serde(default)]
    pub restrict: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub ephemeral_duration: Option<i64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub category: Option<String>,
}

pub struct GroupsRepo<'a> {
    col: Collection<SabwaGroup>,
    _phantom: std::marker::PhantomData<&'a ()>,
}

impl<'a> GroupsRepo<'a> {
    pub fn new(db: &'a Database) -> Self {
        Self {
            col: db.collection::<SabwaGroup>(COLLECTION),
            _phantom: std::marker::PhantomData,
        }
    }

    pub async fn find_by_jid(
        &self,
        session_id: &ObjectId,
        jid: &str,
    ) -> Result<Option<SabwaGroup>> {
        self.col
            .find_one(doc! { "sessionId": session_id, "jid": jid })
            .await
            .context("sabwa_groups.find_by_jid")
    }

    pub async fn upsert(&self, group: &SabwaGroup) -> Result<()> {
        self.col
            .replace_one(
                doc! { "sessionId": &group.session_id, "jid": &group.jid },
                group,
            )
            .upsert(true)
            .await
            .context("sabwa_groups.upsert")?;
        Ok(())
    }

    pub async fn update_participants(
        &self,
        session_id: &ObjectId,
        jid: &str,
        participants: &[Participant],
    ) -> Result<()> {
        let arr = bson::to_bson(participants).context("encode participants")?;
        self.col
            .update_one(
                doc! { "sessionId": session_id, "jid": jid },
                doc! { "$set": { "participants": arr } },
            )
            .await
            .context("sabwa_groups.update_participants")?;
        Ok(())
    }

    pub async fn set_category(
        &self,
        session_id: &ObjectId,
        jid: &str,
        category: Option<&str>,
    ) -> Result<()> {
        let value = match category {
            Some(c) => Bson::String(c.to_string()),
            None => Bson::Null,
        };
        self.col
            .update_one(
                doc! { "sessionId": session_id, "jid": jid },
                doc! { "$set": { "category": value } },
            )
            .await
            .context("sabwa_groups.set_category")?;
        Ok(())
    }

    pub async fn list_by_session(&self, session_id: &ObjectId) -> Result<Vec<SabwaGroup>> {
        let cursor = self
            .col
            .find(doc! { "sessionId": session_id })
            .await
            .context("sabwa_groups.list_by_session")?;
        let out: Vec<SabwaGroup> = cursor.try_collect().await.context("collect groups")?;
        Ok(out)
    }
}

pub fn collection(db: &Database) -> Collection<SabwaGroup> {
    db.collection::<SabwaGroup>(COLLECTION)
}

// ---------------------------------------------------------------------------
// Phase 1 route-compat shims.
// ---------------------------------------------------------------------------

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct GroupParticipantRow {
    pub jid: String,
    pub is_admin: bool,
    pub is_super_admin: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct GroupRow {
    pub jid: String,
    pub subject: String,
    pub description: Option<String>,
    pub creator: Option<String>,
    pub participant_count: u32,
    pub category: Option<String>,
    pub announcement: bool,
    pub restrict: bool,
    pub ephemeral_duration: Option<u32>,
    pub participants: Vec<GroupParticipantRow>,
}

fn group_to_row(g: SabwaGroup) -> GroupRow {
    let participant_count = g.participants.len() as u32;
    GroupRow {
        jid: g.jid,
        subject: g.subject,
        description: g.description,
        creator: g.creator,
        participant_count,
        category: g.category,
        announcement: g.announcement,
        restrict: g.restrict,
        ephemeral_duration: g.ephemeral_duration.and_then(|v| u32::try_from(v).ok()),
        participants: g
            .participants
            .into_iter()
            .map(|p| GroupParticipantRow {
                jid: p.jid,
                is_admin: p.is_admin,
                is_super_admin: p.is_super_admin,
            })
            .collect(),
    }
}

fn parse_oid_loose(id: &str) -> ObjectId {
    ObjectId::parse_str(id).unwrap_or_else(|_| ObjectId::new())
}

pub async fn list(
    db: &Database,
    session_id: &str,
    _category: Option<&str>,
) -> Result<Vec<GroupRow>> {
    let repo = GroupsRepo::new(db);
    let oid = parse_oid_loose(session_id);
    let rows = repo.list_by_session(&oid).await?;
    Ok(rows.into_iter().map(group_to_row).collect())
}

pub async fn get(db: &Database, session_id: &str, jid: &str) -> Result<GroupRow> {
    let repo = GroupsRepo::new(db);
    let oid = parse_oid_loose(session_id);
    let row = repo
        .find_by_jid(&oid, jid)
        .await?
        .ok_or_else(|| anyhow::anyhow!("group not found: {jid}"))?;
    Ok(group_to_row(row))
}

pub async fn set_category(
    db: &Database,
    session_id: &str,
    jid: &str,
    category: &str,
) -> Result<()> {
    let repo = GroupsRepo::new(db);
    let oid = parse_oid_loose(session_id);
    repo.set_category(&oid, jid, Some(category)).await
}

pub async fn get_invite_code(
    db: &Database,
    session_id: &str,
    jid: &str,
) -> Result<Option<String>> {
    let repo = GroupsRepo::new(db);
    let oid = parse_oid_loose(session_id);
    Ok(repo.find_by_jid(&oid, jid).await?.and_then(|g| g.invite_code))
}

File: services/sabwa-node/src/db/groups.ts
/**
 * `sabwa_groups` repository — mirrors the Rust `services/sabwa-engine/src/db/groups.rs`
 * schema 1:1.
 *
 * Stored as `sabwa_groups` collection keyed by `(sessionId, jid)` where `jid`
 * always ends in `@g.us`.
 */

import { ObjectId, type Collection, type Db } from 'mongodb';

export const COLLECTION = 'sabwa_groups';

export interface SabwaParticipant {
  jid: string;
  isAdmin: boolean;
  isSuperAdmin: boolean;
  joinedAt: Date;
}

export interface SabwaGroupDoc {
  _id?: ObjectId;
  projectId: ObjectId;
  sessionId: ObjectId;
  jid: string;
  subject: string;
  description?: string;
  creator?: string;
  createdAt: Date;
  participants: SabwaParticipant[];
  inviteCode?: string;
  announcement?: boolean;
  restrict?: boolean;
  ephemeralDuration?: number;
  category?: string;
}

export interface GroupParticipantRow {
  jid: string;
  isAdmin: boolean;
  isSuperAdmin: boolean;
}

export interface GroupRow {
  jid: string;
  subject: string;
  description: string | null;
  creator: string | null;
  participantCount: number;
  category: string | null;
  announcement: boolean;
  restrict: boolean;
  ephemeralDuration: number | null;
  participants: GroupParticipantRow[];
}

function parseOidLoose(id: string): ObjectId {
  try {
    return new ObjectId(id);
  } catch {
    return new ObjectId();
  }
}

function toRow(g: SabwaGroupDoc): GroupRow {
  const participants = (g.participants ?? []).map((p) => ({
    jid: p.jid,
    isAdmin: p.isAdmin,
    isSuperAdmin: p.isSuperAdmin,
  }));
  return {
    jid: g.jid,
    subject: g.subject,
    description: g.description ?? null,
    creator: g.creator ?? null,
    participantCount: participants.length,
    category: g.category ?? null,
    announcement: g.announcement ?? false,
    restrict: g.restrict ?? false,
    ephemeralDuration: g.ephemeralDuration ?? null,
    participants,
  };
}

export class GroupsRepo {
  private readonly col: Collection<SabwaGroupDoc>;

  constructor(db: Db) {
    this.col = db.collection<SabwaGroupDoc>(COLLECTION);
  }

  async upsert(group: SabwaGroupDoc): Promise<void> {
    await this.col.replaceOne(
      { sessionId: group.sessionId, jid: group.jid },
      group,
      { upsert: true },
    );
  }

  async upsertMany(groups: SabwaGroupDoc[]): Promise<number> {
    if (groups.length === 0) return 0;
    const ops = groups.map((g) => ({
      replaceOne: {
        filter: { sessionId: g.sessionId, jid: g.jid },
        replacement: g,
        upsert: true,
      },
    }));
    const res = await this.col.bulkWrite(ops, { ordered: false });
    return res.upsertedCount + res.modifiedCount;
  }

  async findByJid(sessionId: ObjectId, jid: string): Promise<SabwaGroupDoc | null> {
    return this.col.findOne({ sessionId, jid });
  }

  async listBySession(sessionId: ObjectId, category?: string | null): Promise<SabwaGroupDoc[]> {
    const filter: Record<string, unknown> = { sessionId };
    if (category) filter.category = category;
    return this.col.find(filter).toArray();
  }

  async updateParticipants(
    sessionId: ObjectId,
    jid: string,
    participants: SabwaParticipant[],
  ): Promise<void> {
    await this.col.updateOne({ sessionId, jid }, { $set: { participants } });
  }

  async patchSubject(sessionId: ObjectId, jid: string, subject: string): Promise<void> {
    await this.col.updateOne({ sessionId, jid }, { $set: { subject } });
  }

  async patchDescription(sessionId: ObjectId, jid: string, description: string): Promise<void> {
    await this.col.updateOne({ sessionId, jid }, { $set: { description } });
  }

  async setInviteCode(sessionId: ObjectId, jid: string, inviteCode: string): Promise<void> {
    await this.col.updateOne({ sessionId, jid }, { $set: { inviteCode } });
  }

  async setCategory(sessionId: ObjectId, jid: string, category: string | null): Promise<void> {
    if (category === null) {
      await this.col.updateOne({ sessionId, jid }, { $unset: { category: '' } });
    } else {
      await this.col.updateOne({ sessionId, jid }, { $set: { category } });
    }
  }

  /** Merge participant changes (add/remove/promote/demote). */
  async applyParticipantOp(
    sessionId: ObjectId,
    jid: string,
    op: 'add' | 'remove' | 'promote' | 'demote',
    jids: string[],
  ): Promise<void> {
    const existing = await this.findByJid(sessionId, jid);
    if (!existing) return;
    const map = new Map(existing.participants?.map((p) => [p.jid, p]) ?? []);
    const now = new Date();
    if (op === 'add') {
      for (const j of jids) {
        if (!map.has(j)) {
          map.set(j, { jid: j, isAdmin: false, isSuperAdmin: false, joinedAt: now });
        }
      }
    } else if (op === 'remove') {
      for (const j of jids) map.delete(j);
    } else if (op === 'promote') {
      for (const j of jids) {
        const cur = map.get(j) ?? { jid: j, isAdmin: false, isSuperAdmin: false, joinedAt: now };
        map.set(j, { ...cur, isAdmin: true });
      }
    } else if (op === 'demote') {
      for (const j of jids) {
        const cur = map.get(j);
        if (cur) map.set(j, { ...cur, isAdmin: false, isSuperAdmin: false });
      }
    }
    await this.updateParticipants(sessionId, jid, Array.from(map.values()));
  }
}

// ---------------------------------------------------------------------------
// Route-compat helpers
// ---------------------------------------------------------------------------

export async function list(
  db: Db,
  sessionId: string,
  category?: string | null,
): Promise<GroupRow[]> {
  const repo = new GroupsRepo(db);
  const oid = parseOidLoose(sessionId);
  const rows = await repo.listBySession(oid, category ?? null);
  return rows.map(toRow);
}

export async function get(db: Db, sessionId: string, jid: string): Promise<GroupRow | null> {
  const repo = new GroupsRepo(db);
  const oid = parseOidLoose(sessionId);
  const row = await repo.findByJid(oid, jid);
  return row ? toRow(row) : null;
}

export async function getInviteCode(
  db: Db,
  sessionId: string,
  jid: string,
): Promise<string | null> {
  const repo = new GroupsRepo(db);
  const oid = parseOidLoose(sessionId);
  const row = await repo.findByJid(oid, jid);
  return row?.inviteCode ?? null;
}

export { parseOidLoose };
